feat: publicly export and document the zip64 threshold constants

cosmicexplorer opened the pull request:
See https://github.com/zip-rs/zip-old/pull/402. Closes #60.

cosmicexplorer commented:
It looks like the added test is failing:
```bash
failures:

---- src/spec.rs - spec::ZIP64_BYTES_THR (line 24) stdout ----
Test executable failed (exit status: 101).

stderr:
thread 'main' panicked at src/spec.rs:18:2:
assertion failed: zip.start_file("one.dat", options).is_err()
```

I'll take a look into why this behavior changed since [zip-rs#402](https://github.com/zip-rs/zip-old/pull/402) and try to bisect the failure.

cosmicexplorer commented:
This may be able to use the work from #67.

Pr0methean commented on src/spec.rs:
```diff
@@ -11,7 +13,47 @@ pub(crate) const CENTRAL_DIRECTORY_END_SIGNATURE: u32 = 0x06054b50;
+///
+/// let big_len: usize = (zip::ZIP64_BYTES_THR as usize) + 1;
+/// let big_buf = vec![0u8; big_len];
+/// zip.start_file("zero.dat", options)?;
```
This is wrong: we don't determine until `write_central_and_footer` whether the central directory is ZIP64 or not.

Pr0methean replied:
See https://github.com/zip-rs/zip2/blob/103f1eccc44fb8ad5c2331a342826d0249c64011/src/write.rs#L1289.

Pr0methean commented on src/spec.rs:
```diff
@@ -11,7 +13,47 @@ pub(crate) const CENTRAL_DIRECTORY_END_SIGNATURE: u32 = 0x06054b50;
+/// // This is too big!
+/// assert!(zip.write_all(&big_buf[..]).is_err());
+/// // Attempting to write anything further to the same zip will fail.
+/// assert!(zip.start_file("one.dat", options).is_err());
```
This is wrong; we don't decide until writing the CDE whether it's ZIP64-style or not. See https://github.com/zip-rs/zip2/pull/79/files/2e9ed0d5ba66b93c1e1543c2fdbed0bb60c0e9c7#r1589632794.

cosmicexplorer replied:
Thank you so much!!! Will work on this now.

Pr0methean commented on src/spec.rs:
```diff
@@ -1,4 +1,5 @@
 use crate::result::{ZipError, ZipResult};
+use crate::write::{FileOptions, ZipWriter};
```
Imports that are only used in doc comments should be either guarded by `#[cfg(doc)]` or, as I favor, replaced with fully-qualified names at the point of use.

cosmicexplorer replied:
Done!

Pr0methean reviewed (changes requested):
Are you still planning to work on this PR, or have you folded it into one of the bigger ones?

cosmicexplorer commented:
@Pr0methean this should be ready to review -- I hadn't gotten the hang of the nifty new behavior to rewind any file entries upon write or other errors before, but the test is working now. Fantastic functionality, so incredibly useful.

Pr0methean commented:
@cosmicexplorer Sorry if the merge for this PR is delayed. I've been experimenting with new ways to generate and improve the fuzz corpus, and have found more panic-causing bugs tonight than I would've expected in a year -- and that's *before* cross-pollinating the corpora from different restarts!

Pr0methean commented:
Of course, if you're able to fix any of these bugs yourself in separate PRs, I'll merge them very gratefully.

cosmicexplorer commented:
Rebased and fixed conflicts; sorry for dropping this one! really want to get a new version of the parallel extraction PR #72 going (since that one is much more straightforward/less error-prone than the rest of #193, and I believe also much more requested by users), but will absolutely be trying to keep an eye out for bugfixes of that sort. I really like reading the code you've added to this crate and fixing bugs is a joy. Please especially feel free to ping me in the future if there's anything in particular you might like an extra hand with.

Pr0methean commented on src/spec.rs:
```diff
@@ -93,8 +94,55 @@ impl ExtraFieldMagic {
 }
 
-/// This should be equal to `0xFFFFFFFF`.
+/// The number of bytes necessary to allocate a zip64 record for an individual file.
+///
+/// If a file larger than this threshold attempts to be written, and
```
```suggestion
/// The file size at which a ZIP64 record becomes necessary.
///
/// If you attempt to write a file this size or larger, compressed or uncompressed, and
```
I believe a ZIP64 record is needed when the file is *exactly* this size, just to clarify that it's not strictly larger, and I recently changed the code to reflect that.

cosmicexplorer replied:
Hmmm, so I just tried testing this, and it seems that at exactly this size the write does *not* fail. See diff:
```diff
diff --git a/src/spec.rs b/src/spec.rs
index 1148cd7ca..9525fef41 100755
--- a/src/spec.rs
+++ b/src/spec.rs
@@ -113,14 +113,11 @@ impl ExtraFieldMagic {
 /// // Writing an extremely large file for this test is faster without compression.
 /// let options = SimpleFileOptions::default().compression_method(zip::CompressionMethod::Stored);
 ///
-/// let big_len: usize = (zip::ZIP64_BYTES_THR as usize) + 1;
+/// let big_len: usize = zip::ZIP64_BYTES_THR as usize;
 /// let big_buf = vec![0u8; big_len];
 /// zip.start_file("zero.dat", options)?;
 /// // This is too big!
-/// let res = zip.write_all(&big_buf[..]).err().unwrap();
-/// assert_eq!(res.kind(), io::ErrorKind::Other);
-/// let description = format!("{}", &res);
-/// assert_eq!(description, "Large file option has not been set");
+/// let res = zip.write_all(&big_buf[..]).unwrap();
 /// // Attempting to write anything further to the same zip will still succeed, but the previous
 /// // failing entry has been removed.
 /// zip.start_file("one.dat", options)?;
```

Just changing `big_len` to equal `ZIP64_BYTES_THR` causes the doctest to fail, because the `.err().unwrap()` call fails. When I change `.err().unwrap()` to just `.unwrap()` and remove the checking of the error description (as in the diff), the zip is successfully created, with both file entries. Do you think this is wrong?

I have made the other changes you suggested to the docstring, however!

cosmicexplorer commented:
CI is failing on a separate change, I'll push a quick fix for that now in a separate PR.

cosmicexplorer commented:
Ping @Pr0methean!

Pr0methean commented:
This has a merge conflict, which I'll try to resolve once the queue is empty.

cosmicexplorer commented:
The merge conflict mysteriously disappeared!

Pr0methean commented:
It also mysteriously reappeared - apparently the conflict was with another PR that got merged today but had been queued before.

Pr0methean commented:
The cause of the conflict is that spec.rs no longer imports `core::mem::align_of`.

cosmicexplorer commented:
Fixed!
